drivers/adc/ad4858/ad4858.c

```diff
@@ -0,0 +1,1069 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include "ad4858.h"
+#include "no_os_delay.h"
+#include "no_os_print_log.h"
+
+/**
+ * @brief Write device register.
+ * @param dev- The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_val - The data to be written.
+ * @return 0 in case of success, negative error code otherwise.
+ * @note Multibyte write with streaming mode is supported by default.
+ */
+int ad4858_reg_write(struct ad4858_dev *dev, uint32_t reg_addr,
+		     uint32_t reg_val)
+{
+	int32_t ret;
+	uint32_t addr;
+	uint8_t cnt = 0;
+	uint8_t reg_size = AD4858_LEN(reg_addr);
+	uint8_t buf[6] = { 0 };
+
+	if (!dev)
+		return -EINVAL;
+
+	if (!dev->addr_ascension_enable) {
+		addr = AD4858_ADDR(reg_addr) + reg_size - 1;
+	} else {
+		addr = AD4858_ADDR(reg_addr);
+	}
+
+	buf[cnt++] = (uint8_t)(addr >> 8);
+	buf[cnt++] = (uint8_t)addr;
+
+	if (dev->big_endian) {
+		if (dev->addr_ascension_enable)
+			no_os_memswap64((uint8_t *)(&reg_val + 1) - reg_size, reg_size, reg_size);
+		memcpy(&buf[cnt], (uint8_t *)(&reg_val + 1) - reg_size, reg_size);
+	} else {
+		if (!dev->addr_ascension_enable)
+			no_os_memswap64(&reg_val, reg_size, reg_size);
+		memcpy(&buf[cnt], &reg_val, reg_size);
+	}
+
+	/* Add the register data size to calculate an actual length of SPI frame */
+	cnt += reg_size;
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buf, cnt);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+/**
+ * @brief Read device register.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_val - The data read from the register.
+ * @return 0 in case of success, negative error code otherwise.
+ * @note Multibyte read with streaming mode is supported by default.
+ */
+int ad4858_reg_read(struct ad4858_dev *dev, uint32_t reg_addr,
+		    uint32_t *reg_val)
+{
+	int32_t ret;
+	uint32_t addr;
+	uint8_t cnt = 0;
+	uint8_t dindx = 0;
+	uint8_t reg_size = AD4858_LEN(reg_addr);
+	uint8_t buf[6] = { 0 };
+
+	if (!dev || !reg_val)
+		return -EINVAL;
+
+	if (!dev->addr_ascension_enable) {
+		addr = AD4858_ADDR(reg_addr) + reg_size - 1;
+	} else {
+		addr = AD4858_ADDR(reg_addr);
+	}
+
+	/* Store the address (16-bit instruction phase with R/W=1) */
+	buf[cnt++] = (uint8_t)(AD4858_REG_RD_BIT_MSK | (addr >> 8));
+	buf[cnt++] = (uint8_t)addr;
+
+	/* Save the data phase start index */
+	dindx = cnt;
+
+	/* Add the register data size to calculate an actual length of SPI frame */
+	cnt += reg_size;
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buf, cnt);
+	if (ret)
+		return ret;
+
+	*reg_val = 0;
```

Comment:
is this explicit initialization required?

Reply:
No I guess, will remove this